Feat request all user datato42 server

joejaeyoung opened the pull request:
1. Access Token 발급을 위한 코드 값을 받아
2. 42api를 호출하여 전체 유저정보를 받아와
3. string 형식으로 리턴합니다.

## 추가 개선 필요 사항
- code값이 들어오지 않았을 경우 처리
- token값이 들어오지 않았을 경우 처리
- 받은 데이터 정보 무결성 검증
- 현 controller 역할 분담시키기

joejaeyoung committed:
feat : Data load from 42api

joejaeyoung committed:
feat:Load all user data without insert code

joejaeyoung committed:
fix: 리턴값 변경 및 코드 발급 실패시 오류 리턴

joonwan commented on src/main/java/com/ggs/STAT_TurboFetch/api/DataLoadController.java:
```diff
@@ -0,0 +1,100 @@
+import com.ggs.STAT_TurboFetch.api.dto.Oauth2TokenRequestDto;
+import com.ggs.STAT_TurboFetch.api.dto.Oauth2TokenResponseDto;
+import lombok.RequiredArgsConstructor;
+import org.springframework.http.*;
```
필요한 부분만 import 하면 좋을것 같습니다.

joejaeyoung replied:
```

import org.springframework.http.HttpEntity;
import org.springframework.http.HttpHeaders;
import org.springframework.http.HttpMethod;

```

헤더 명시해 두겠습니다!

joonwan commented on src/main/java/com/ggs/STAT_TurboFetch/api/DataLoadController.java:
```diff
@@ -0,0 +1,100 @@
+public class DataLoadController {
+    private final TokenConfig tokenConfig;
+
+    @GetMapping("/init")
```
endpoint 명은 어떤 것을 초기화 하는지 명시적으로 보여주면 좋을 것 같습니다. 
사용자의 목록을 반환하기 위한 end point기 때문에 `/users` 라고 해도 괜찮을 것 같습니다.

joejaeyoung replied:
엔드포인트 명명 규칙 확인했고 수정하겠습니다!

joonwan commented on src/main/java/com/ggs/STAT_TurboFetch/api/DataLoadController.java:
```diff
@@ -0,0 +1,100 @@
+    private final TokenConfig tokenConfig;
+
+    @GetMapping("/init")
+    public String returnAllUsers(
```
반환값을 한번 감싸는 객체가 있었으면 좋겠습니다. SRR Project 에 ApiResponse 를 참고하면 좋을 것 같습니다.
또한 String 으로 모든 문자열을 결합하여 반환하기 보다는 List 에 각 데이터를 담아서 이 List 를 응답 전용 객체에 담는 것을 추천합니다.

joejaeyoung replied:
ApiResponse로 한번 감싸서 반환하겠습니다.

joonwan commented on src/main/java/com/ggs/STAT_TurboFetch/api/DataLoadController.java:
```diff
@@ -0,0 +1,100 @@
+        1. 일단 .yml로 관리
+        2. request parameter로 관리
+         */
+        System.out.println("=== code ===");
```
확인을 하기 위해 출력하는 것은 `System.out.println()` 메서드 대신 @slf4j 를 통해 `log.info()` 같은 메서드를 사용하는 것을 추천합니다. 이는 불필요한 연산을 줄여 성능적인 측면에서 더 좋습니다.

joejaeyoung replied:
기능 | System.out.println | 로깅 프레임워크 (e.g., Logback, Log4j)
-- | -- | --
I/O 처리 | 동기화된 블로킹 I/O 사용 | 비동기 I/O 지원 가능
스레드 안전성 | 동기화로 인해 병목 발생 가능 | 비동기 처리 및 병목 감소
로그 수준 필터링 | 지원하지 않음 | 로그 레벨 기반 필터링 지원
출력 대상 다양성 | 콘솔 출력만 가능 | 파일, 원격 서버 등 다양한 대상 지원
지연 로깅 | 지원하지 않음 | 문자열 처리 비용 최적화
성능 | 대량 로그 시 성능 저하 | 고성능 (최적화된 I/O 처리)
구현 복잡성 | 간단함 | 설정 필요 (초기 설정 시간 소요)

성능차이 확인 후, log() 메소드로 변경하겠습니다.
간단한 디버그 문으로 info레벨로 설정하겠습니다.

joonwan commented on src/main/java/com/ggs/STAT_TurboFetch/api/DataLoadController.java:
```diff
@@ -0,0 +1,100 @@
+        - 데이터 파싱
+        - 시간 개선
+         */
+        final String CAMPUS_USER_URI = "https://api.intra.42.fr/v2/cursus_users?filter[campus_id]=69&filter[cursus_id]=21&page=";
```
상수는 지역변수로 잡기 보다는 클래스 레벨의 상수로 뽑는 것을 추천합니다. 

```java

public class DataLoadController {

    private static final String CAMPUS_USER_URI = "https://api.intra.42.fr/v2/cursus_users?filter[campus_id]=69&filter[cursus_id]=21&page=";
    private final TokenConfig tokenConfig;

    ...
}
```

이런식으로 말이죠

joejaeyoung replied:
변광했습니다.

joonwan commented on src/main/java/com/ggs/STAT_TurboFetch/api/DataLoadController.java:
```diff
@@ -0,0 +1,100 @@
+            ArrayList<HashMap<String, Object>> body = response.getBody();
+
+            if (body.isEmpty()) break;
+            allUsers += body;
```
List 에 응답 받은 문자열을 담아주면 좋겠습니다.

만약 문자열을 반복하여 결합하는 로직이 있을 경우 String 인스턴스들을 직접 더하는 것은 비효율 적입니다. String 은 불변 객체이기 때문에 더하는 연산이 발생할 때 마다 새로운 객체가 생성되어 성능이 안좋아집니다. 대신 `StringBuilder` class 를 사용하는 것을 추천합니다.

joejaeyoung replied:
가변적 문자열:

String은 불변(immutable) 객체이므로 문자열을 변경하면 항상 새로운 객체가 생성됩니다.
반면, StringBuilder는 동일 객체 내에서 문자열을 변경하므로 메모리 효율성이 높습니다.
싱글 스레드 환경:

StringBuilder는 스레드 안전(Thread-safe) 하지 않습니다.
멀티스레드 환경에서는 StringBuffer를 사용하세요.
성능:

문자열의 빈번한 변경(추가, 삭제, 수정 등)이 필요한 경우 StringBuilder가 String보다 빠릅니다.

내용 확인했고, 
`List<String> allUsers = new ArrayList<>();`
로 변경하겠습니다!

joonwan reviewed (changes requested):
고생하셨습니다! 자세한 내용은 코멘트를 확인해 주세요!!

joejaeyoung committed:
1. 필요 라이브러리만 호출되게 필요없는 라이브러리 import 삭제
2. 엔드포인트 이름 /init -> /users 로 변경
3. sout대신 더 성능 좋은 log()메소드 사용
4. 상수는 클래스 레벨로 변경
5. 응답받은 데이터 String -> List<String>으로 변경
6. 최종 반환값 String -> ApiResponse<T>로 한번 감싸줘서 반환

joejaeyoung committed:
fix: 기존 controller역할 분담 및 Access Token 발급 시 401 예외 커스텀 예외로 변경

joonwan reviewed (approved):
lgtm👍🏿
